### PR TITLE
fix: Add sample notebook for tracing and debugging agents in Reasoning Engine

### PR DESCRIPTION
**Pull request by koverholt**
# Description

This PR adds a sample notebook for tracing, debugging, and 

- [x] Follow the [`CONTRIBUTING` Guide](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/CONTRIBUTING.md).
- [x] You are listed as the author in your notebook or README file.
  - [x] Your account is listed in [`CODEOWNERS`](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/.github/CODEOWNERS) for the file(s).
- [x] Make your Pull Request title in the <https://www.conventionalcommits.org/> specification.
- [x] Ensure the tests and linter pass (Run `nox -s format` from the repository root to format).
- [x] Appropriate docs were updated (if necessary)

### CODE REVIEW

**Comment by koverholt** [comment]
Still a work in progress, but most of the content and flow is in place. I have a few more details to work through, including additional narrative text, updating the intro section, and switching out tools for a different example & use case. Will continue working on this in this PR.

File: gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb [hunk 1]
```diff
@@ -0,0 +1,1304 @@
...
+    "# Debugging and Optimizing Agents: A Guide to Tracing in Reasoning Engine\n",
+    "\n",
+    "<table align=\"left\">\n",
+    "  <td style=\"text-align: center\">\n",
+    "    <a href=\"https://colab.research.google.com/github/GoogleCloudPlatform/generative-ai/blob/main/gemini/reasoning-engine/tracing_agents_in_reasoning_engine\">\n",
+    "      <img src=\"https://cloud.google.com/ml-engine/images/colab-logo-32px.png\" alt=\"Google Colaboratory logo\"><br> Open in Colab\n",
+    "    </a>\n",
+    "  </td>\n",
+    "  <td style=\"text-align: center\">\n",
+    "    <a href=\"https://console.cloud.google.com/vertex-ai/colab/import/https:%2F%2Fraw.githubusercontent.com%2FGoogleCloudPlatform%2Fgenerative-ai%2Fmain%2Fgemini%2Freasoning-engine.ipynb\">\n",
```
Comment (holtskinner) [inline gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb:41]: Make sure to update this link

File: gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb [hunk 2]
```diff
@@ -0,0 +1,1304 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "84f0f73a0f76"
+   },
+   "source": [
+    "| | |\n",
+    "|-|-|\n",
+    "|Author(s) | [Kristopher Overholt](https://github.com/koverholt)"
```
Comment (holtskinner) [inline gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb:66]:
```suggestion
    "| Author(s) | [Kristopher Overholt](https://github.com/koverholt) |"
```

File: gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb [hunk 3]
```diff
@@ -0,0 +1,1304 @@
...
+    "        params={\"from\": currency_from, \"to\": currency_to},\n",
+    "    )\n",
+    "    return response.json()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Define agent and enable traceing"
```
Comment (holtskinner) [inline gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb:370]:
```suggestion
    "### Define agent and enable tracing"
```

File: gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb [hunk 4]
```diff
@@ -0,0 +1,1304 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 47,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "project_id: \"koverholt-devrel-355716\"\n",
```
Comment (holtskinner) [inline gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb:472]: Remove this project ID

File: gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb [hunk 5]
```diff
@@ -0,0 +1,1304 @@
...
+      "Using bucket koverholt-devrel-355716-bucket\n",
+      "Writing to gs://koverholt-devrel-355716-bucket/reasoning_engine/reasoning_engine.pkl\n",
+      "Writing to gs://koverholt-devrel-355716-bucket/reasoning_engine/requirements.txt\n",
+      "Creating in-memory tarfile of extra_packages\n",
+      "Writing to gs://koverholt-devrel-355716-bucket/reasoning_engine/dependencies.tar.gz\n",
+      "Creating ReasoningEngine\n",
+      "Create ReasoningEngine backing LRO: projects/964731510884/locations/us-central1/reasoningEngines/1586287415624990720/operations/1605235058988285952\n",
+      "ReasoningEngine created. Resource name: projects/964731510884/locations/us-central1/reasoningEngines/1586287415624990720\n",
+      "To use this ReasoningEngine in another session:\n",
+      "reasoning_engine = vertexai.preview.reasoning_engines.ReasoningEngine('projects/964731510884/locations/us-central1/reasoningEngines/1586287415624990720')\n"
```
Comment (holtskinner) [inline gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb:592]: Remove project IDs

File: gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb [hunk 6]
```diff
@@ -0,0 +1,1304 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 69,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pandas as pd\n",
+    "from google.cloud import trace_v1\n",
```
Comment (holtskinner) [inline gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb:322]:
```suggestion
    "from google.cloud import trace_v1 as trace\n",
```

Just a personal pet peeve about our libraries, if this doesn't cause any name collisions, can you change references to `trace_v1` to just `trace`

File: gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb [hunk 7]
```diff
@@ -0,0 +1,1304 @@
...
+    "from vertexai.preview import reasoning_engines\n",
+    "from vertexai.reasoning_engines._reasoning_engines import _utils"
```
Comment (holtskinner) [inline gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb:324]: This is a little weird that you need to import a protected module. I wonder if there's something we can change to the SDK to make this easier?

**Review by holtskinner — APPROVED** [review]

File: gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb (hunk 1 above)
Comment (koverholt) [reply gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb]: Done!

File: gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb (hunk 2 above)
Comment (koverholt) [reply gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb:66]: Ahh, I fixed this in the `notebook_template`. Thanks!

File: gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb (hunk 3 above)
Comment (koverholt) [reply gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb]: Fixed

File: gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb (hunk 6 above)
Comment (koverholt) [reply gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb]: Thank you! This was irking me and I couldn't think of how to make it better. Your solution is 😎 . Fixed.

File: gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb (hunk 7 above)
Comment (koverholt) [reply gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb:324]: Agreed - I believe the next version of the SDK will make these private modules and methods public. So this should only be needed for the initial / current SDK version that supports tracing.

File: gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb (hunk 4 above)
Comment (koverholt) [reply gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb]: Done

File: gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb (hunk 5 above)
Comment (koverholt) [reply gemini/reasoning-engine/tracing_agents_in_reasoning_engine.ipynb]: Done

**Pull request merged**